src/Engine/ProtoCore/DSASM/InstructionSet.cs

```diff
@@ -362,6 +342,258 @@ public bool IsReferenceType
         }
         #endregion
 
+        private void Check(bool cond, string errorMessage)
+        {
+            if (!cond)
+                throw new InvalidStackValueTypeException(errorMessage);
+        }
+
+        #region Type-dependent value extractor
+        public int VariableIndex
+        {
+            get
+            {
+                Check(IsVariableIndex, "The type of StackValue is not VariableIndex");
+                return (int)opdata;
+            }
+        }
+
+        public int FunctionIndex
+        {
+            get
+            {
+                Check(IsFunctionIndex, "The type of StackValue is not FunctionIndex");
+                return (int)opdata;
+            }
+        }
+
+        public int MemberVariableIndex
+        {
+            get
+            {
+                Check(IsMemberVariableIndex, "The type of StackValue is not MemberVariableIndex");
+                return (int)opdata;
+            }
+        }
+
+        public int StaticVariableIndex
+        {
+            get
+            {
+                Check(IsStaticVariableIndex, "The type of StackValue is not StaticVariableIndex");
+                return (int)opdata;
+            }
+        }
+
+        public int SymbolIndex
+        {
+            get
+            {
+                Check(IsVariableIndex || IsStaticVariableIndex || IsMemberVariableIndex, "It is not a symbol type");
+                return (int)opdata;
+            }
+        }
+
+        public int ClassIndex
+        {
+            get
+            {
+                Check(IsClassIndex, "The Type of StackValue is not ClassIndex");
+                return (int)opdata;
+            }
+        }
+
+        public long IntegerValue
+        {
+            get
+            {
+                Check(IsInteger, "The Type of StackValue is not Integer");
+                return (int)opdata;
+            }
+        }
+
+        public long CharValue
+        {
+            get
+            {
+                Check(IsChar, "The Type of StackValue is not Char");
+                return (int)opdata;
+            }
+        }
```

Comment:
Why do both `IntegerValue` and `CharValue` need to return `long` when it is always truncated to `int`? I see few callers below are casting it back to `int`, so perhaps we can rethink about it?